File: app/src/main/java/net/emilla/util/Chars.java
package net.emilla.util;

public enum Chars {
    ;

    public static boolean isLineSeparator(int ch) {
        return ch == '\n' || ch == '\r';
    }

    public static boolean isNonLineSpace(int ch) {
        return !isLineSeparator(ch) && Character.isWhitespace(ch);
    }

    public static boolean differentLetters(char a, char b) {
        return compareIgnoreCase(a, b) != 0;
    }

    private static int compareIgnoreCase(char a, char b) {
        if (a != b && Character.toUpperCase(a) != Character.toUpperCase(b)) {
            a = Character.toLowerCase(a);
            b = Character.toLowerCase(b);
            if (a != b) return a - b;
        }

        return 0;
    }

    public static boolean isSignOrDigit(char ch) {
        return isSign(ch) || Character.isDigit(ch);
    }

    public static boolean isSignOrNumberChar(char ch) {
        return isSign(ch) || isNumberChar(ch);
    }

    public static boolean isNumberChar(char ch) {
        return ch == '.' || Character.isDigit(ch);
    }

    private static boolean isSign(char ch) {
        return switch (ch) {
            case '+', '-' -> true;
            default -> false;
        };
    }

}

File: app/src/main/java/net/emilla/util/Chars.kt
@file:JvmName("Chars")

package net.emilla.util

fun Char.notSpace() = !isWhitespace()

fun Char.isNonLineSpace() = when (this) {
    '\n', '\r' -> false
    else -> isWhitespace()
}

@JvmName("differentLetters")
fun Char.differentLetter(c: Char): Boolean {
    return compareToIgnoreCase(c) != 0
}

@JvmName("compareIgnoreCase")
fun Char.compareToIgnoreCase(c: Char): Int {
    if (this != c && uppercaseChar() != c.uppercaseChar()) {
        val a = lowercaseChar()
        val b = c.lowercaseChar()
        if (a != b) return a - b
    }

    return 0
}

fun Char.isSignOrDigit() = isDigit() || isSign()
fun Char.isSignOrNumberChar() = isNumberChar() || isSign()

fun Char.isDigit() = this in '0'..'9'

fun Char.isNumberChar() = this == '.' || this in '0'..'9'

fun Char.isSign() = when (this) {
    '+', '-' -> true
    else -> false
}
